### DISCUSSION TITLE
load a specific Berry script failed sometimes

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20195] [Q&A] [answered]
In this [discussion](https://github.com/arendst/Tasmota/discussions/20173#discussioncomment-7779928), I noticed a mistake.
>`BRY: failed to load '/controll.be' (syntax_error - /controll.be:13: 'Licht' undeclared (first use in this function))
false`

I initially identified it as a presumed syntax error at the end. But that's not the case.
Perhaps it's also related to this [problem](https://github.com/arendst/Tasmota/discussions/20171#discussioncomment-7792047)?

I have a script (controll.be) that includes another script at the beginning using load("SensorAktor.be"). It only defines classes and it used to work. Suddenly, sometimes this file is no longer loaded. if i load both files seperatly or copy the content from `SensorAktor.be` to `Controll.be` it will work. I have no glue what needs to come back working with `load("SensorAktor.be"). Then i try randomly deleting or editing this file or restarting and reuploading until it works. That feels bad.

the picture belows show this behavior. Restart Berry VM returns an error.
loading both files by hand, works fine.

![grafik](https://github.com/arendst/Tasmota/assets/16272493/884d7827-646b-4a8d-aa9e-a3fa270cc529)

Any ideas?

**Comment by ChrisVonFrosta** [comment]
just after my post i did only a `restart 1` by  console and it  starts perfectly... i do not know...

**Comment by s-hadinger** [comment]
Berry needs to know the globals at compile time. `load` is executed at runtime, hence any global declared in the sub-file are not visible at compile time. Normally you would favor using `import` and never rely on any global variable, unless the ones from `import`

**Comment by ChrisVonFrosta** [comment]
thx, but then, why it not happens all the time?
As i said. the next restart did it normal.
the `SensorAktor.be` has global vars but they are used as defines and only used in `SensorAktor.be`.

`import` means as module?

**Reply by sfromis** [reply]
In that specific case, as a workaround you could have a `var CCTLicht` (the class name) before the `load` to allow the compiler to process the whole block of code, without depending on what's declared in the loaded code, which will only become available when the code is executed.

Defining a class results in a global variable named after the class, and housing the object which is the compiled code for the class. Therefore, this name is not available to the compiler when it "only" can compile the `load` as one function to be executed at run time.

Arguably, using `load` is not well-suited as part of a program, where using a `module` and `import` indeed makes for less quirks to work around.

**Reply by ChrisVonFrosta** [reply]
Thank you! Sorry i was away for a longt time.
That means that i need to create a helper function for each function of the module that i want to use outside the module?

example:

`module.be`
```
var module_name=module("module_name")

class aktor
  def init(var_init)
     ...
  end

  def do_somthing()
  ...
  end

end

module_name.init = def (name, var_init)
  name=aktor(var_init)
def

module_name.do_something = def ()
  ...
end

return module_name
```

that wouldn't pretty convinent.

**Reply by sfromis** [reply]
No, you **need** no separate helper functions, whatever you return from the module becomes the value in the variable named after the `import`. If what you return is a class instance, all methods of the class are thus available. If you return a class, static methods become available.

But a lot depends on how you want to manage instance creation, and also how your `do_something` is related to instances. If it is part of an instance, it will be available if that instance becomes visible outside of the module. And if it not part of an instance, you can have the helper function code directly in `module_name.do_something = def ()`

Anyway, your "example" does not demonstrate what you mean, as it does not make sense no matter what would be in the `...` parts. Specifically, your `name=aktor(var_init)` creates an instance, but is not doing anything with it. That's again about how your instances are to be managed, without that it is not practical to suggest what to do.

**Reply by ChrisVonFrosta** [reply]
i am so sorry for my confusing example and bad english!

i got your point. it works with one class but i stuck again with more classes.

How to return multiple classes from the module??
Do I need to pack all the classes into an array or a list beforehand?

i tried this stupid thing: `return Aktor,CCTLicht`
and  got the error:
`BRY: failed to run compiled code 'syntax_error' - /test1.be:377: unexpected symbol near ','`

**Reply by sfromis** [reply]
I'd say that if you want to return multiple classes, what you do is just not a good case for a `module`. The intention is not to see a `module` as some sort of "include", but instead to encapsulate a feature with a clearly defined API, and then hiding how it works internally.

That API might include factory methods to return instances of a class, but if you want to return a class as such, you are no longer doing an API, as the module has no control over what is done.

Still, if you set a module variable to the class object, you have external access to a class from the module. No need for arrays or other weird things.... You can mess around many ways if what you want is not a "real" module.

Anytime you make an assignment to `module_name.xxxx` that is simply a variable which can refer to all sorts of different things, not just the common use case of functions to be called.

(For context: When you make a function definition like `def myfunc(name) print('hello',name) end` you are really creating a variable named `myfunc`, and then that variable contains the function, in the same way as defining a `class` is also simply creating a variable of that name).

**Reply by ChrisVonFrosta** [reply]
Thank you very much, also for your patience. I started with a single script file. However, it grew large, and the part that needs to be changed often for testing and also in live operation is almost at the bottom: the creation of objects for consumers and switches.

So, I would like to have the definitions of classes and methods in one file and the object creation and call logic (trigger rules) in another, just like with include files.

How would you fragment this? I will add the structure of my script; maybe that will make it more understandable. Everything before "object creation" should be separated. I couldn't think of anything better than a module. Using load("file.be") did not work.

`pseudo script`

```

# Aktor Classes
# ################################################

class Aktor
    #...
end


# Subklasse 
class CCTLicht : Aktor
    
    def init()      # init sub-class 
        #... some variables
    end
    
    
    def off()       # light off
        #... some more stuff
    end

    def toggle()    # light toggle
        #... some more stuff
    end

    #... some more stuff

end


# Sensor Classes
# ################################################

class Sensor
    #...

    static def triggerNo(input_number, input_value)     # Pcalled by parse_trigger_msg()  
    
    end
end

# Subklassen        
class Taster : Sensor

    def trigger(trigger)        # called by Sensor.triggerNo()
        #... some more stuff
    end

    def execute(push_cnt)       # called by self.trigger(trigger)and execute the action(s) for this button
        #... some more stuff
    end

    #... some more stuff
end



# Object creation
# ################################################


var l_wohnzimmer     = CCTLicht(...)   # create CCTLight Object

var t_wohnzimmer    =  Taster(...)   # create button Object



# ################################################
# Called by trigger with button number and -value
def  parse_trigger_msg(value, trigger, msg)

    #print("value:", value)
    #print("trigger:", trigger)

    var string_value = map.tostring(value)
    #var string_trigger = map.tostring(trigger)

    var input_block     =   int(trigger[8])
    var input_number    =   int(string_value[3])
    var input_value     =   int(string_value[7])

    print("\n\n***********************\n button cycle Start \n***********************\nn")

    print("input_block:", input_block)          #   PCF8574-2, PCF8574-3
    print("input_number:", input_number)        #   Port
    print("input_value:", input_value)          #   Value: True/false

    # var cnt_buttons = list.size(buttons)

    if int(input_block) == 3
        input_number += 8
        print("new input_number:", input_number)        #   Port
    end
        
    sensor.triggerNo(input_number, input_value)

end


tasmota.add_rule("PCF8574-2_INP", parse_trigger_msg)
tasmota.add_rule("PCF8574-3_INP", parse_trigger_msg)
```

**Reply by sfromis** [reply]
If your **only** problem is that you want to first do a `load("file.be")` and then have more code in `autoexec.be`, the simplest would be to just do that `load` in `preinit.be` which will be executed before `autoexec.be`.

There is nothing which is similar to a C-style include, as in a precompiler constructing a longer source before starting to compile. With Berry it is a single-pass compiler, meaning that you cannot depend on definitions only available after having executed some of the code.

It does not make sense to try to use a `module` as merely a way to load code without structuring it as an actual `module`, which can include ways of exposing what you'd like to expose. Anything which you do not expose is by design "private" to the module.

**Reply by ChrisVonFrosta** [reply]
so you do berry scripting in one file everytime? you never splitting a large berry script?

Since almost nothing is private in the classes or the module, I wanted to expose everything. The only reason for splitting is overview and handling.

> meaning that you cannot depend on definitions only available after having executed some of the code.
Am I doing that? 
my `autoexec.be` loads `controll.be`.
My main script `controll.be` should create the instances after loading/importing the file/module `SensorAktor.be` with the defintions of  classes/objects and methods.. Has code been executed at this point? 

I see 3 Options.
1.  the whole script in one file (and had to update a large scrip file for adding buttons or lights or changing the button behavior).
2.  (miss)using the concept of modules.
3.  Using a `preinit.be` file (did not understand your explanation because `autoexec.be`).

But I don't want to be a nuisance. If it seems hopeless with me... ;)
I had assumed that splitting a script would be a common endeavor even in Berry.

**Reply by sfromis** [reply]
I certainly do use building block `module` components sometimes, that is not the same as merely trying to break a longer script into files. Instead is to create a enclosed feature with a clearly defined API, designed for reuse. Basically designing to not have it be "one large script" with everything globally visible.

But again, what you are trying to do is the opposite of a `module`, as you want to "expose everything". A main purpose of a `module` is to only expose carefully selected ways of using the `module`.

I'm not trying to say that this is a "problem", merely that if this is what you want, the `preinit.be` approach is probably more like what you want. Using `preinit.be` works very similar to `autoexec.be`, the difference is that `preinit.be` is executed at an earlier stage in the boot process.

Sure, you *can* use modules for something like what you want, as it is possible to also have a module expose a class directly, in the way I already I mentioned earlier. Whatever rocks your boat, also when it is not "real modules" :grinning:

**Reply by sfromis** [reply, marked as answer]
Example of exposing whole classes, and nothing else, the way I told you about. `class_container.be`:
```
var class_container = module('class_container')

class class1
  def init()
    print('class1 created')
  end
end

class class2
  def init()
    print('class2 created')
  end
end

class_container.class1 = class1
class_container.class2 = class2

return class_container
```

And then:
```
import class_container as cc
var ci1 = cc.class1()
var ci2 = cc.class2()
```

Of course, you could also expose what you may like to have of helper methods or whatever. Exposing from a module is really simple, as a var can be basically any object, including classes and functions.

**Reply by ChrisVonFrosta** [reply]
This is exactly what I've been looking for. I wanted to try this as a last resort, but each time I ended up with no remote access to the ESP32.

Thank you so much for pointing this out, for your confirmation, and your patience!

**Comment by ChrisVonFrosta** [comment]
> Anyway, your "example" does not demonstrate what you mean, as it does not make sense no matter what would be in the ... parts. Specifically, your name=aktor(var_init) creates an instance, but is not doing anything with it. That's again about how your instances are to be managed, without that it is not practical to suggest what to do.

Yes, I understand. I simplified a lot because, in the actual code, a multitude of variables are passed during initialization, and these have nothing to do with my actual problem: "how to use/create modules correctly." Is there any documentation or an example for this somewhere? Should I post the entire (possibly cluttered) code next time, or is it better to post a simplified but functional minimal example?

**Reply by sfromis** [reply]
See https://github.com/berry-lang/berry/wiki/Chapter-8#83-how-to-package-a-module

I'm not trying to suggest what you "should post", but if you want to ask about doing something, example code which does not demonstrate what you try to do can't lead to point out where it does not work. But such code will be seen as an attempt to illustrate the idea.
